File: C#/SocketClient.cs
using System.Collections;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Net;
using System.Threading.Tasks;
using System.Text;
using System;

/// <summary>
/// Ñlass implements the behavior of a socket client. Uses to connect to server, send and receive messages
/// </summary>
public class SocketClient
{
    /// <summary>
    /// Server IP address
    /// </summary>
    public string serverIP = "127.0.0.1";

    /// <summary>
    /// Port binded by the server
    /// </summary>
    public int serverPort = 7777;

    /// <summary>
    /// Name of client
    /// </summary>
    public string clientName = "";

    /// <summary>
    /// True if client connected to server
    /// </summary>
    public bool isConnected = false;

    /// <summary>
    /// Uses to contain task for listening for messages
    /// </summary>
    protected Task messageReciver;

    /// <summary>
    /// Socket client
    /// </summary>
    protected Socket client;

    public SocketClient() { }
    public SocketClient(string serverIP, int serverPort, string clientName)
    {
        this.serverIP = serverIP;
        this.serverPort = serverPort;
        this.clientName = clientName;
    }

    /// <summary>
    /// Called when client connected to server
    /// </summary>
    public virtual void OnConnected() { }

    /// <summary>
    /// Called when client disconnected to server
    /// </summary>
    public virtual void OnDisconnected() { }

    /// <summary>
    /// Called when message delivered to server
    /// </summary>
    /// <param name="message">Message text</param>
    public virtual void OnMessageDelivered() { }

    /// <summary>
    /// Called when recived message from server
    /// </summary>
    /// <param name="message">Message text</param>
    public virtual void OnMessageRecived(string message) { }

    /// <summary>
    /// Connects client to server
    /// </summary>
    public async void ConnectAsync()
    {
        if (isConnected)
        {
            RSMALogger.Logger.Log($"{clientName}: Ñlient is already connected to server");
            return;
        }
        IPAddress ipAddress = IPAddress.Parse(serverIP);

        IPEndPoint ipEndPoint = new(ipAddress, serverPort);

        client = new(ipEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

        try 
        {
            await client.ConnectAsync(ipEndPoint);
        }
        catch(Exception ex)
        {
            RSMALogger.Logger.Log($"{clientName}: Connection error: The destination computer rejected the connection request. Make sure that the server is running and available for connection");
            return;
        }

        RSMALogger.Logger.Log($"{clientName}: Waiting for server responce");

        while (true)
        {
            string message = $"{clientName}<|CR|><|EOM|>";
            var messageBytes = Encoding.UTF8.GetBytes(message);
            try
            {
                _ = await client.SendAsync(messageBytes, SocketFlags.None);
            }
            catch (Exception ex)
            {
                RSMALogger.Logger.Log($"{clientName}: Connection error: The destination computer rejected the connection request. Make sure that the server is running and available for connection");
                return;
            }
            var buffer = new byte[1024];
            int received = 0;
            try
            {
                received = await client.ReceiveAsync(buffer, SocketFlags.None);
            }
            catch (Exception ex)
            {
                RSMALogger.Logger.Log($"{clientName}: Connection error: The destination computer rejected the connection request. Try change client/user name");
                return;
            }
            var response = Encoding.UTF8.GetString(buffer, 0, received);

            if (response == $"<|CR|><|ACK|>")
            {
                RSMALogger.Logger.Log($"{clientName}: successfully connect to server.");
                isConnected = true;
                break;
            }
        }

        messageReciver = Task.Run(() => ListenToMessageAsync());

        OnConnected();
    }
    /// <summary>
    /// Sends message to server
    /// </summary>
    /// <param name="message">Message text</param>
    public async void SendMessageAsync(string message)
    {
        if (!isConnected)
        {
            RSMALogger.Logger.Log($"{clientName}: client is NOT connected to server");
            return;
        }
        message = $"{clientName}<|M|>" + message + "<|EOM|>";
        var messageBytes = Encoding.UTF8.GetBytes(message);

        try
        {
            _ = await client.SendAsync(messageBytes, SocketFlags.None);
        }
        catch(Exception ex)
        {
            RSMALogger.Logger.Log($"{clientName}: Listener error: The remote computer has terminated an existing connection. Make sure that the server is running and available for connection");
            isConnected = false;
            return;
        } 
    }
    /// <summary>
    /// Disconnects cleint from server
    /// </summary>
    public async void DisconnectAsync()
    {
        if (!isConnected)
        {
            RSMALogger.Logger.Log($"{clientName}: Ñlient is NOT connected to server");
            return;
        }
        string message = $"{clientName}<|DR|><|EOM|>";
        var messageBytes = Encoding.UTF8.GetBytes(message);

        try
        {
            _ = await client.SendAsync(messageBytes, SocketFlags.None);
        }
        catch (Exception ex)
        {
            RSMALogger.Logger.Log($"{clientName}: Listener error: The remote computer has terminated an existing connection. Make sure that the server is running and available for connection");
            isConnected = false;
            return;
        }
    }
    /// <summary>
    /// Listening to messages
    /// </summary>
    protected async void ListenToMessageAsync()
    {
        while (true)
        {
            byte[] buffer = new byte[1024];
            int receivedBites = 0;
            try
            {
                receivedBites = await client.ReceiveAsync(buffer, SocketFlags.None);
            }
            catch
            {
                RSMALogger.Logger.Log($"{clientName}: Listener error: The remote computer has terminated an existing connection. Make sure that the server is running and available for connection");
                isConnected = false;
                return;
            }
            string receivedText = Encoding.UTF8.GetString(buffer, 0, receivedBites);

            if (receivedText.IndexOf("<|EOM|>") > -1 || receivedText.IndexOf("<|ACK|>") > -1)
            {
                string[] messages = receivedText.Split("<|EOM|>");
                string[] acks = receivedText.Split("<|ACK|>");

                if(messages.Length > 1)
                {
                    foreach (var message in messages)
                    {
                        if (message.IndexOf("<|M|>") > -1)
                        {
                            string text = message.Split("<|M|>")[1];

                            string ackMessage = $"<|M|><|ACK|>";
                            byte[] echoBytes = Encoding.UTF8.GetBytes(ackMessage);
                            await client.SendAsync(echoBytes, 0);

                            RSMALogger.Logger.Log($"{clientName}: recived message from server");

                            OnMessageRecived(text);
                        }
                    }
                }

                if (acks.Length > 1)
                {
                    foreach (var ack in acks)
                    {
                        if (ack == "<|M|>")
                        {
                            RSMALogger.Logger.Log($"{clientName}: Message delivered");
                            OnMessageDelivered();
                        }
                        else if (ack == "<|DR|>")
                        {
                            try
                            {
                                messageReciver.Dispose();
                                client.Shutdown(SocketShutdown.Both);
                            }
                            catch (Exception ex)
                            {
                                RSMALogger.Logger.Log(ex.Message);
                                return;
                            }
                            finally
                            {
                                client.Close();
                                isConnected = false;
                                RSMALogger.Logger.Log($"{clientName}: Disconnected");
                                OnDisconnected();
                            }
                            return;
                        }
                    }
                }
            }
        }
    }
}

// massage struct: target/clientName<separator>messageText<|EOM|>

// separators: <|EOM|> - end of message;
// <|M|> - message separator
// <|CR|> - client connection request separator
// <|DR|> - client disconnection request separator'
// <|ACK|> - ack separator

// ack struct: <separator><|ACK|>

File: Python/socket_client.py
import socket
import threading

class SocketClient:

    server_IP: str
    """Server IP address"""

    server_port: int
    """Port binded by the server"""

    client_name: str
    """Name of client"""

    is_connected: bool
    """True if client connected to server"""

    _client: socket
    """Socket client"""

    _message_reciver: threading.Thread
    """Uses to contain task for listening for messages"""

    logger: bool
    """If true, logging client messages"""

    def __init__(self, server_IP: str, server_port: int, client_name: str):
        self.server_IP = server_IP
        self.server_port = server_port
        self.client_name = client_name
        self.is_connected = False
        self._message_reciver = None
        self.logger = False
    
    def _listen_to_message_async(self):
        """Listening to messages"""
        while(True):
            buffer: bytes
            try:
                buffer = self._client.recv(1024)
            except:
                print(f"{self.client_name}: Connection error: The destination computer rejected the connection request. Try change client/user name")
                return
            
            received_text = buffer.decode()

            if(received_text.find("<|EOM|>") > -1 or received_text.find("<|ACK|>") > -1):
                messages = received_text.split("<|EOM|>")
                acks = received_text.split("<|ACK|>")

                if(len(messages) > 1):
                    for message in messages:
                        if(message.find("<|M|>") > -1):
                            text = message.split("<|M|>")[1]

                            ack_message = "<|M|><|ACK|>"
                            echo_buffer = ack_message.encode()
                            self._client.send(echo_buffer)

                            if(self.logger):
                                print(f"{self.client_name}: recived message from server")

                            self.on_message_recived(text)

                if(len(acks) > 1):
                    for ack in acks:
                        if(ack == "<|M|>"):
                            if(self.logger):
                                print(f"{self.client_name}: Message delivered")
                            self.on_message_delivered()
                        elif(ack == "<|DR|>"):
                            try:
                                self._message_reciver.join()
                                self._client.close()
                            except Exception as e:
                                print(e)
                                return
                            finally:
                                self._client.close()
                                self.is_connected = False
                                print(f"{self.client_name}: Disconnected")
                                self.on_disconnected
                            return

    def connect(self):
        """Connects client to server"""
        if(self.is_connected):
            print(f"{self.client_name}: Сlient is already connected to server")
            return
        
        self._client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        try:
            self._client.connect((self.server_IP, self.server_port))
        except:
            print(f"{self.client_name}: Connection error: The destination computer rejected the connection request. Make sure that the server is running and available for connection")
            return 
        if(self.logger):
            print(f"{self.client_name}: Waiting for server responce")

        message = f"{self.client_name}<|CR|><|EOM|>"
        message_bytes = message.encode()

        try:
            self._client.send(message_bytes)
        except:
            print(f"{self.client_name}: Connection error: The destination computer rejected the connection request. Make sure that the server is running and available for connection")
            return
        
        buffer: bytes
        try:
            buffer = self._client.recv(1024)
        except:
            print(f"{self.client_name}: Connection error: The destination computer rejected the connection request. Try change client/user name")
            return
        
        response = buffer.decode()

        if(response == "<|CR|><|ACK|>"):
            print(f"{self.client_name}: successfully connect to server.")
            self.is_connected = True
        else:
            self._client.close

        if(self._message_reciver == None):
            _message_reciver = threading.Thread(target=self._listen_to_message_async, daemon=True)
            _message_reciver.start()

        self.on_connected()

    def send_message(self, message_text: str):
        """Sends message to server"""
        if(not self.is_connected):
            print(f"{self.client_name}: client is NOT connected to server")
            return
        
        message = f"{self.client_name}<|M|>" + message_text + "<|EOM|>"
        message_bytes = message.encode()

        try:
            self._client.send(message_bytes)
        except:
            print(f"{self.client_name}: Listener error: The remote computer has terminated an existing connection. Make sure that the server is running and available for connection")
            self.is_connected = False
            return
    
    def disconnect(self):
        """Disconnects cleint from server"""
        if(not self.is_connected):
            print(f"{self.client_name}: client is NOT connected to server")
            return
        
        message = f"{self.client_name}<|DR|><|EOM|>"
        message_bytes = message.encode()

        try:
            self._client.send(message_bytes)
        except:
            print(f"{self.client_name}: Listener error: The remote computer has terminated an existing connection. Make sure that the server is running and available for connection")
            self.is_connected = False
            return

    def on_message_recived(self, text: str):
        """Called when recived message from server"""

    def on_message_delivered(self):
        """Called when message delivered to server"""

    def on_connected(self):
        """Called when client connected to server"""
    
    def on_disconnected(self):
        """Called when client disconnected to server"""

